### PR TITLE
Generating client

### PR DESCRIPTION
**Pull request by acbeni**
Yet another huge PR, sorry @katmatt for that 😢 

Before merging the pr please make sure it doesn't break anything since it contains clean ups for old code generators.

To make the new generator work you have to setup the test runner to gradle test runner in intellij otherwise it doesn't recognise the extensions.

### CODE REVIEW

File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/StringExtension.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import io.vrap.rmf.codegen.common.processor.annotations.ModelExtension;
+import org.apache.commons.lang3.StringUtils;
+
+import java.text.CharacterIterator;
+import java.text.StringCharacterIterator;
+
+@ModelExtension(extend = String.class)
+public class StringExtension {
+
+    private static final Converter<String, String> classNameMapper = CaseFormat.LOWER_HYPHEN.converterTo(CaseFormat.UPPER_CAMEL);
```
Comment (jenschude) [inline common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/StringExtension.java]: Take a look at io.vrap.rmf.raml.model.util.StringCaseFormat which takes any cased input and converts it to the target format.

File: annotations/src/main/java/io/vrap/rmf/codegen/common/processor/processor/ModelExtensionProcessor.java [hunk 2]
```diff
@@ -169,10 +172,12 @@ CodeBlock getApplyCodeBlock(TypeElement typeElement, ClassName extendedTypeName)
                 ));
 
         codeBlockBuilder.add("  default:\n")
-                .add("      result = null;\n")
+                .addStatement("       LOGGER.debug(\"property '{}' not found in type {}\", trimmedMethodName, $T.class )",getExtensionMapperName(typeElement))
```
Comment (katmatt) [inline annotations/src/main/java/io/vrap/rmf/codegen/common/processor/processor/ModelExtensionProcessor.java]: Nice that you added some logging here 👍

File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/core/JavaSTFileSwitch.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class JavaSTFileSwitch extends ComposedSwitch<STGroupFile> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JavaSTFileSwitch.class);
+
+    private Map<String, STGroupFile> cache = new HashMap<>();
```
Comment (katmatt) [inline common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/core/JavaSTFileSwitch.java]: I'm not 100% sure, but I thought that stringtemplate already provides a cache. But let's keep your code as it is and we can still see if we can improve this further in the future.

File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/core/JavaSTFileSwitch.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        rmfModelAdaptor = new RmfModelAdaptor(Object.class, extensionMappers);
+        addSwitch(new JavaSTFileTypesSwitch());
+    }
+
+    public STGroupFile getSTFileFor(Object object) {
+        STGroupFile result;
+        if (object instanceof EObject) {
+            result = doSwitch((EObject) object);
+        } else if (object instanceof EObjectsCollection) {
+            if (((EObjectsCollection) object).getSample() instanceof Resource) {
```
Comment (katmatt) [inline common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/core/JavaSTFileSwitch.java]: That's the trick you mentioned. And it's fine for me 😄

File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/DescriptionFacetExtension.java [hunk 5]
```diff
@@ -23,4 +23,9 @@ public void setJavaDocProcessor(JavaDocProcessor javaDocProcessor) {
     public String getDescription(DescriptionFacet descriptionFacet) {
         return javaDocProcessor.markDownToJavaDoc(descriptionFacet);
     }
+
+    @ExtensionMethod
+    public Boolean hasDescription(DescriptionFacet descriptionFacet) {
```
Comment (katmatt) [inline common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/DescriptionFacetExtension.java]: This extension looks very usefule 👍

File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/MethodExtension.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    @ExtensionMethod
+    public AnyType getReturnType(final Method method) {
+        return method.getResponses().stream()
+                .filter(r -> r.getStatusCode().equals("200"))
+                .filter(response -> (response.getBodies() != null) && (!response.getBodies().isEmpty()))
+                .findFirst()
+                .map(method1 -> method1.getBodies().get(0))
+                .map(TypedElement::getType)
+                .orElse(TypesFactoryImpl.eINSTANCE.createNilType());
```
Comment (katmatt) [inline common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/MethodExtension.java]: I replaced your custom type with a proper RMF type here. Looks like I should give you a better introduction into RMF/EMF after your vacation 😄

File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/ObjectTypeExtension.java [hunk 7]
```diff
@@ -64,6 +64,7 @@ public boolean hasSubtypes(ObjectType objectType) {
                 .flatMapIterable(this::getAllClassNames)
                 .filter(s -> !s.startsWith("java.lang"))
                 .filter(s -> !s.startsWith("java.util"))
+                .filter(s -> !s.startsWith(className.packageName()))
```
Comment (katmatt) [inline common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/extensions/ObjectTypeExtension.java]: Importing one self was never a good idea 😉 👍

**Pull request merged**